kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -583,43 +605,65 @@ private static void validateFieldCompatibility(
                   newField.getName(), currentKeyType, newKeyType));
         }
       }
+    } else {
+      for (Class<DataType> clazz : new Class[] {StructType.class, ArrayType.class}) {
+        if (existingField.getDataType().getClass() == clazz
+            && newField.getDataType().getClass() == clazz) {
+          return;
+        }
+      }
+      if (!existingField.getDataType().equivalent(newField.getDataType())) {
+        throw new KernelException(
+            String.format(
+                "Cannot change the type of existing field %s from %s to %s",
+                existingField.getName(), existingField.getDataType(), newField.getDataType()));
+      }
+    }
+  }
 
-      validateFieldCompatibility(
-          existingMapType.getKeyField(), newMapType.getKeyField(), icebergWriterCompatV1Enabled);
-      validateFieldCompatibility(
-          existingMapType.getValueField(),
-          newMapType.getValueField(),
-          icebergWriterCompatV1Enabled);
-    } else if (existingField.getDataType() instanceof ArrayType
-        && newField.getDataType() instanceof ArrayType) {
-      ArrayType existingArrayType = (ArrayType) existingField.getDataType();
-      ArrayType newArrayType = (ArrayType) newField.getDataType();
-
-      validateFieldCompatibility(
-          existingArrayType.getElementField(),
-          newArrayType.getElementField(),
-          icebergWriterCompatV1Enabled);
-    } else if (!existingField.getDataType().equivalent(newField.getDataType())) {
-      throw new KernelException(
-          String.format(
-              "Cannot change the type of existing field %s from %s to %s",
-              existingField.getName(), existingField.getDataType(), newField.getDataType()));
+  private static class SchemaElementId {
```

Comment:
some comments to say what this is used for? and the relation between getPath and id?

Reply:
done.